File: attendance_analyzer.cpp
#include <iostream>
#include <set>
#include <map>
#include <vector>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <algorithm>
using namespace std;

struct Student {
    string id, name;
};

map<string, Student> students;
map<string, set<string>> attendance;

void addStudent() {
    Student s;
    cout << "Student ID: "; cin >> s.id;
    cin.ignore();
    cout << "Student Name: "; getline(cin, s.name);
    students[s.id] = s;
    cout << "Student added.\n";
}

void markAttendance() {
    string date, id;
    cout << "Date (e.g. 2026-09-04): "; cin >> date;
    cout << "Enter present student IDs (0 to finish):\n";
    while (cin >> id && id != "0") {
        if (students.count(id)) attendance[date].insert(id);
        else cout << "Student not found: " << id << "\n";
    }
    cout << "Attendance saved.\n";
}

void showSetOperations() {
    string d1, d2;
    cout << "First date: "; cin >> d1;
    cout << "Second date: "; cin >> d2;

    set<string> a = attendance[d1], b = attendance[d2], result;
    cout << "\nA = " << d1 << "\nB = " << d2 << "\n";

    set_union(a.begin(), a.end(), b.begin(), b.end(), inserter(result, result.begin()));
    cout << "A union B: ";
    for (auto id : result) cout << id << " ";
    cout << "\n";

    result.clear();
    set_intersection(a.begin(), a.end(), b.begin(), b.end(), inserter(result, result.begin()));
    cout << "A intersection B: ";
    for (auto id : result) cout << id << " ";
    cout << "\n";

    result.clear();
    set_difference(a.begin(), a.end(), b.begin(), b.end(), inserter(result, result.begin()));
    cout << "A - B: ";
    for (auto id : result) cout << id << " ";
    cout << "\n";

    result.clear();
    set_symmetric_difference(a.begin(), a.end(), b.begin(), b.end(), inserter(result, result.begin()));
    cout << "Symmetric difference: ";
    for (auto id : result) cout << id << " ";
    cout << "\n";
}

void report() {
    if (attendance.empty()) { cout << "No attendance records.\n"; return; }
    cout << "\nAttendance Report\n";
    cout << left << setw(12) << "ID" << setw(22) << "Name" << "Percentage\n";
    cout << string(48, '-') << "\n";
    int totalDays = attendance.size();
    for (auto &[id, s] : students) {
        int present = 0;
        for (auto &[date, ids] : attendance)
            if (ids.count(id)) present++;
        double pct = totalDays ? present * 100.0 / totalDays : 0;
        cout << left << setw(12) << id << setw(22) << s.name
             << fixed << setprecision(2) << pct << "%\n";
    }
}

void saveData() {
    ofstream out("attendance_data.txt");
    for (auto &[id, s] : students)
        out << "STUDENT|" << s.id << "|" << s.name << "\n";
    for (auto &[date, ids] : attendance)
        for (const auto &id : ids)
            out << "ATT|" << date << "|" << id << "\n";
    cout << "Data saved to attendance_data.txt\n";
}

void loadData() {
    ifstream in("attendance_data.txt");
    string line;
    while (getline(in, line)) {
        stringstream ss(line);
        string type, a, b;
        getline(ss, type, '|'); getline(ss, a, '|'); getline(ss, b);
        if (type == "STUDENT") students[a] = {a, b};
        else if (type == "ATT") attendance[a].insert(b);
    }
}

int main() {
    loadData();
    int choice;
    do {
        cout << "\n=== Student Attendance Analyzer Using Set Theory ===\n";
        cout << "1. Add Student\n2. Mark Attendance\n3. Set Theory Analysis\n";
        cout << "4. Attendance Report\n5. Save Data\n0. Exit\nChoice: ";
        cin >> choice;
        switch (choice) {
            case 1: addStudent(); break;
            case 2: markAttendance(); break;
            case 3: showSetOperations(); break;
            case 4: report(); break;
            case 5: saveData(); break;
            case 0: saveData(); cout << "Goodbye!\n"; break;
            default: cout << "Invalid choice.\n";
        }
    } while (choice != 0);
    return 0;
}

File: attendance_analyzer.py
import json
from pathlib import Path

DATA_FILE = Path("attendance_data.json")
students = {}
attendance = {}

def save():
    DATA_FILE.write_text(json.dumps({
        "students": students,
        "attendance": {d: list(ids) for d, ids in attendance.items()}
    }, indent=2))
    print("Data saved.")

def load():
    global students, attendance
    if DATA_FILE.exists():
        data = json.loads(DATA_FILE.read_text())
        students = data.get("students", {})
        attendance = {d: set(ids) for d, ids in data.get("attendance", {}).items()}

def add_student():
    sid = input("Student ID: ").strip()
    name = input("Student Name: ").strip()
    students[sid] = name
    print("Student added.")

def mark_attendance():
    date = input("Date (YYYY-MM-DD): ").strip()
    ids = input("Present student IDs separated by spaces: ").split()
    valid = {sid for sid in ids if sid in students}
    attendance[date] = valid
    print(f"{len(valid)} students marked present.")

def set_analysis():
    d1 = input("First date: ").strip()
    d2 = input("Second date: ").strip()
    a, b = attendance.get(d1, set()), attendance.get(d2, set())
    print("\nA =", a)
    print("B =", b)
    print("A union B =", a | b)
    print("A intersection B =", a & b)
    print("A - B =", a - b)
    print("B - A =", b - a)
    print("Symmetric difference =", a ^ b)

def report():
    days = len(attendance)
    print("\n=== Attendance Report ===")
    if not days:
        print("No attendance records.")
        return
    for sid, name in students.items():
        present = sum(sid in ids for ids in attendance.values())
        pct = present / days * 100
        status = "LOW ATTENDANCE" if pct < 75 else "OK"
        print(f"{sid:12} {name:25} {pct:6.2f}%  {status}")

def main():
    load()
    while True:
        print("\n=== Student Attendance Analyzer Using Set Theory ===")
        print("1. Add Student")
        print("2. Mark Attendance")
        print("3. Set Theory Analysis")
        print("4. Attendance Report")
        print("5. Save Data")
        print("0. Exit")
        choice = input("Choice: ").strip()
        if choice == "1": add_student()
        elif choice == "2": mark_attendance()
        elif choice == "3": set_analysis()
        elif choice == "4": report()
        elif choice == "5": save()
        elif choice == "0":
            save()
            print("Goodbye!")
            break
        else: print("Invalid choice.")

if __name__ == "__main__":
    main()
